### PR TITLE
Initial MVP implementation

### LINKED ISSUE #23: Start initial MVP using existing engine features

### PR DESCRIPTION
**Pull request by JoXW100**
Fixes #23 
Implements the MVP with the current engine features.

- Added gameobjects to render scene from project proposal
- Added `Renderer::DrawLine()`
- Added `Renderer::DrawSprite()`

image

### CODE REVIEW

**Comment by caspernorrbin** [comment]
I expanded the MVP to include basic networking with a server and client, as well as a game controller and a network controller. This also starts to implement the game itself as it was required for everything else. The game controller controls the game and holds other game objects, like ships and the network controller. It can currently read and store the state it receives from the server and send actions back.

To aid in testing, every 5 server turns (5 seconds at the moment), the game controller sends an action to the server. All classes also have debug member variables that control printing debug messages. These should be moved and refined once the logger is implemented.

A lot of improvements are still to be made and the implementation is still WIP, so any suggested changes are welcome.

File: src/mvp/commontypes.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    static const uint8_t Attack = 2;
+};
+
+struct ShipInfo {
+    uint16_t Damage;
+    uint16_t Health;
+    uint16_t Cost;
+    uint16_t textureOffsetX;
+    uint16_t textureOffsetY;
+};
```
Comment (joelsiks) [inline src/mvp/commontypes.hpp:36]: Why are these structs with static data instead of enums that inherit from a type? Like:

```c++
struct NetworkMessageTypes : uint16_t {
    PlayerReady = 0;
    ReadyConfirmation;
    GameStart;
    // ...
}
```

File: src/mvp/main.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            GameData::engine->MakeUIElement<objects::IconifiedButton>(
+                "buyShip" + std::to_string(shipType), 0,
+                std::to_string(ship.Cost), Vector2(GameData::CellSize),
+                Color::WHITE, Color::BLACK, atlas,
+                Vector2(ship.textureOffsetX, ship.textureOffsetY));
+        buyShipButton->SetDisplayPosition(UI::DisplayPosition::LowerLeft);
+        buyShipButton->onClick += [gameManager, shipType](void *, auto) {
+            gameManager->BuyShip(shipType);
+        };
+    }
```
Comment (joelsiks) [inline src/mvp/main.cpp:89]: There are a lot of stuff happening here, perhaps we can move out some things to their own functions, like `CreateGameBoard(...)`

File: src/mvp/main.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    GameData::engine->StartGameLoop();
+
+    return EXIT_SUCCESS;
+}
+
+// #define TRANSPARENT _TRANSPARENT
```
Comment (joelsiks) [inline src/mvp/main.cpp:96]: Huh?

File: src/mvp/main.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+#include "objects/Grid.hpp"
+#include "objects/IconifiedButton.hpp"
+#include "objects/Quad.hpp"
+#include "objects/Ship.hpp"
+#include "shared.hpp"
+#include <SDL_vulkan.h>
+#include <VK2D/Constants.h>
+#include <VK2D/VK2D.h>
+
+#undef TRANSPARENT
```
Comment (joelsiks) [inline src/mvp/main.cpp:14]: Why is this necessary? A comment perhaps?

File: src/mvp/main.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+#include "objects/Background.hpp"
+#include "objects/GameManager.hpp"
+#include "objects/Grid.hpp"
+#include "objects/IconifiedButton.hpp"
+#include "objects/Quad.hpp"
+#include "objects/Ship.hpp"
+#include "shared.hpp"
+#include <SDL_vulkan.h>
+#include <VK2D/Constants.h>
+#include <VK2D/VK2D.h>
```
Comment (joelsiks) [inline src/mvp/main.cpp:12]: These three are not needed here.

File: src/mvp/main.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#include "Engine.hpp"
+#include "UI/TextElement.hpp"
+#include "objects/Background.hpp"
+#include "objects/GameManager.hpp"
+#include "objects/Grid.hpp"
+#include "objects/IconifiedButton.hpp"
+#include "objects/Quad.hpp"
+#include "objects/Ship.hpp"
+#include "shared.hpp"
```
Comment (joelsiks) [inline src/mvp/main.cpp:9]: I would separate the engine includes and the MVP includes by a space, i.e:

```c++
#include "Engine.hpp"
#include "UI/TextElement.hpp"

#include "objects/Background.hpp"
#include "objects/GameManager.hpp"
#include "objects/Grid.hpp"
#include "objects/IconifiedButton.hpp"
#include "objects/Quad.hpp"
#include "objects/Ship.hpp"
#include "shared.hpp"
```

File: src/engine/UI/Button.hpp [hunk 7]
```diff
@@ -15,14 +15,14 @@ class Button : public Element {
     Button(const std::string &name, float order, const std::string &text,
            const Vector2 &size, const Color &bgColor, const Color &fgColor);
 
-    void Render(const Texture &font) override;
+    virtual void Render(const Texture &font) override;
```
Comment (joelsiks) [inline src/engine/UI/Button.hpp:18]: Why?

**Review by joelsiks — COMMENTED** [review]
Perhaps the game should be paused, closed or exited when one of the players disconnect.

There is a lot of code to look through here so I have probably missed some stuff, but I've commented on the things that I find stand out.

**Review by joelsiks — APPROVED** [review]
